#### Quiz - INTRODUCTION TO PYTHON FOR MACHINE LEARNING

In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
url='https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv'
data=pd.read_csv(url)
data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

## Q.1 


##### If you’re given two lists:

A = [1,2,3,4,5,6], B = [13, 21, 34]

The task is to create a list with the elements of A and B in a single dimension with output:

A_B = [1,2,3,4,5,6,13, 21, 34]

Which of the following option is the best way to create this list?



Ans- None of the above

# Q. 2

How do you create an identity matrix in python?

In [3]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

Ans- np.identity(3) 

# Q.3

Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [7]:
data.groupby(['fuel_type_code_pudl']).mean()['fuel_cost_per_unit_burned'].sort_values()

fuel_type_code_pudl
gas          13.659397
other        18.253856
waste        19.518122
coal         67.421830
oil         168.877086
nuclear    4955.157002
Name: fuel_cost_per_unit_burned, dtype: float64

Ans- Gas

# Q.4


What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [8]:
np.round(data['fuel_mmbtu_per_unit'].std(),2)

10.6

In [9]:
np.round(np.quantile(data['fuel_mmbtu_per_unit'],0.75),2)

17.01

Ans- 10.60 and 17.01

# Q.5

What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [11]:
np.round(data['fuel_qty_burned'].skew(),2)

15.85

In [12]:
np.round(data['fuel_qty_burned'].kurtosis(),2)

651.37

Ans- 15.85 and 651.37

# Q.6

Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places?

(Enter answer in the format: Feature: xxx, Total: xxx, Percent: xxx)

In [14]:
data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [16]:
data.shape

(29523, 11)

In [17]:
np.round(((180/29523)*100),3)

0.61

Ans - Feature: fuel_unit,Total:180,Percent:0.61

# Q.7

The feature with missing values falls under what category? What missing value imputation technique would you use?

In [18]:
data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

Ans- Categorical and mode imputation

# Q.8

Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [23]:
data.corr()['fuel_cost_per_unit_burned'].sort_values(ascending=False)

fuel_cost_per_unit_burned       1.000000
report_year                     0.013599
fuel_cost_per_unit_delivered    0.011007
fuel_cost_per_mmbtu            -0.000437
fuel_mmbtu_per_unit            -0.010034
fuel_qty_burned                -0.018535
utility_id_ferc1               -0.037863
Name: fuel_cost_per_unit_burned, dtype: float64

Ans-fuel_qty_burned and fuel_mmbtu_per_unit

# Q.9

For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [24]:
coal_df=data[data['fuel_type_code_pudl']=='coal']

In [26]:
coal_df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
5  f1_fuel_1994_12_2_0_13                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
5   e. c. gaston-unit 5                coal       ton        2124933.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
5               23.922                      44.24   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
5                         43.25                1.850

In [28]:
coal_df.groupby(['report_year']).sum()['fuel_cost_per_unit_burned']

report_year
1994     14984.572
1995     14571.785
1996     12694.803
1997     43742.178
1998     11902.597
1999     16303.468
2000    100428.831
2001     15109.012
2002      9171.164
2003     30954.110
2004      9632.063
2005     11107.182
2006     12402.684
2007     12965.171
2008     14331.348
2009     45554.858
2010     88627.745
2011     17001.015
2012     17546.097
2013     17629.860
2014     15412.197
2015     13338.235
2016     11100.925
2017     10254.284
2018      9488.197
Name: fuel_cost_per_unit_burned, dtype: float64

In [30]:
(11902.597-14984.572)/14984.572

-0.20567654518260517

Ans-  -21%

# Q.10

Which year has the highest average fuel cost per unit delivered?

In [34]:
data.groupby(['report_year']).mean()['fuel_cost_per_unit_delivered'].sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

Ans- 1997

## Thank You!!